File: 0017_letterCombinationsOfAPhoneNumber/main.cpp
#include <vector>
#include <string>

class Solution {
public:
    std::vector<std::string> letterCombinations(std::string digits) {
        if(digits.empty())
            return {};
        std::string map[10] = {
            "",    //0
            "",    //1
            "abc", //2
            "def", //3
            "ghi", //4
            "jkl", //5
            "mno", //6
            "pqrs",//7
            "tuv", //8
            "wxyz" //9
        };

        std::vector<std::string> res({""});

        for(int i = 0; i < digits.size(); i++) {
            std::vector<std::string> tmp;
            std::string s = map[digits[i] - '0'];
            for(int j = 0; j < s.size(); j++) {
                for(int k = 0; k < res.size(); k++) {
                    tmp.push_back(res[k] + s[j]);
                }
            }
            res.swap(tmp);
        }
        return res;
    }

    std::vector<std::string> letterCombinations(std::string digits) {
        if(digits.empty())
            return {};
        std::string map[10] = {
            "",    //0
            "",    //1
            "abc", //2
            "def", //3
            "ghi", //4
            "jkl", //5
            "mno", //6
            "pqrs",//7
            "tuv", //8
            "wxyz" //9
        };

        std::vector<std::string> res;
        std::queue<std::string, std::list<std::string>> myQueue({""});

        while(!myQueue.empty()) {
            std::string curString = myQueue.front();
            myQueue.pop();

            if(curString.size() == digits.size())
                res.push_back(curString);
            else {
                std::string mapString = map[digits[curString.size()] - '0'];
                for(int i = 0; i < mapString.size(); i++)
                    myQueue.push(curString + mapString[i]);
            }
        }
        return res;

    }

};

File: 0017_letterCombinationsOfAPhoneNumber/main.py
from typing import List

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        res = []
        digitToChar = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "qprs",
            "8": "tuv",
            "9": "wxyz",
        }

        def backtrack(startIdx, curStr):
            if len(curStr) == len(digits):
                res.append(curStr)
                return 
            
            for c in digitToChar[digits[startIdx]]:
                backtrack(startIdx+1, curStr + c)

        if digits:
            backtrack(0, "")

        return res